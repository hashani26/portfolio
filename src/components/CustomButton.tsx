import React from "react";
import { Button } from "antd";
import { ButtonProps } from "antd/lib/button";

export const CustomButton = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => (
    <>
      <Button data-testid="button" ref={ref} {...props}>
        {props.children}
      </Button>
    </>
  )
);
