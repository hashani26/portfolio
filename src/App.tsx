import React, { createRef, useState, useRef, RefAttributes } from "react";
import { Row, Col, Card, Table, Typography, Input } from "antd";
import { ColumnProps } from "antd/lib/table";
import { DeleteOutlined } from "@ant-design/icons";
import _ from "lodash";
import "antd/dist/antd.less";
import { CustomButton } from "./components/CustomButton";
import Button, { ButtonProps } from "antd/lib/button";
import "../src/styles/index.less";

const { Title } = Typography;

interface TaskList {
  key: number;
  task: string;
  status: string;
  remove: React.ReactNode;
}

function App() {
  const [taskName, setTaskName] = useState<string>("");
  const [data, setData] = useState<TaskList[]>([]);
  // const statusBtn = React.useRef() as React.MutableRefObject<HTMLButtonElement>;
  const statusBtn = React.useRef<HTMLButtonElement>();
  console.log(
    "ðŸš€ ~ file: App.tsx ~ line 23 ~ App ~ statusBtn",
    statusBtn.current
  );

  const columns: ColumnProps<TaskList>[] = [
    {
      title: "Task",
      dataIndex: "task",
    },
    {
      title: "Status",
      dataIndex: "status",
      render: (value, record, index) => (
        <CustomButton
          ref={createRef()}
          type="primary"
          onClick={() => {
            data[index].status = value === "To do" ? "Done" : "To do";
            setData(_.cloneDeep(data));
          }}
        >
          {value}
        </CustomButton>
      ),
    },
    {
      title: "",
      dataIndex: "remove",
      render: (value, record, index) => (
        <DeleteOutlined
          onClick={() => {
            data.splice(index, 1);
            setData(_.cloneDeep(data));
          }}
        />
      ),
    },
  ];

  function recordText(event: React.FormEvent<HTMLInputElement>) {
    setTaskName(event.currentTarget.value);
  }

  function addTask() {
    data.push({
      key: data.length + 1,
      task: taskName,
      status: "To do",
      remove: "",
    });
    setData(_.cloneDeep(data));
    setTaskName("");
  }

  return (
    <>
      <Row>
        <Col xs={{ span: 24 }} lg={{ span: 12, offset: 6 }}>
          <Title className={"text-center mt-16"}>To Do Application</Title>
          <Card bordered={true}>
            <Row gutter={[16, 16]}>
              <Col lg={{ span: 20 }} xs={{ span: 24 }}>
                <Input value={taskName} onChange={recordText} />
              </Col>
              <Col lg={{ span: 4 }} xs={{ span: 24 }}>
                <CustomButton
                  type="primary"
                  disabled={_.isEmpty(taskName)}
                  onClick={addTask}
                >
                  ADD TASK
                </CustomButton>
              </Col>
            </Row>
            <Table
              scroll={{ y: "50vh" }}
              pagination={false}
              columns={columns}
              dataSource={data}
            />
          </Card>
        </Col>
      </Row>
    </>
  );
}

export default App;
