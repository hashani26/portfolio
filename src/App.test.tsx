import "./matchMedia.js";
import React from "react";
import { render, screen, cleanup } from "@testing-library/react";
import App from "./App";
import ReactDOM from "react-dom";
import { CustomButton } from "./components/CustomButton";
import renderer from "react-test-renderer";

afterEach(cleanup);

test("renders app title To Do Application", () => {
  render(<App />);
  const linkElement = screen.getByText(/To Do Application/i);
  expect(linkElement).toBeInTheDocument();
}); //do a test to find if the title is same as To Do Application

it("Renders button component without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<CustomButton />, div);
});

it("renders button correctly", () => {
  const { getByTestId } = render(<CustomButton>Add Task</CustomButton>);
  expect(getByTestId("button")).toHaveTextContent("Add Task");
});

it("renders button correctly", () => {
  const { getByTestId } = render(<CustomButton>Test</CustomButton>);
  expect(getByTestId("button")).toHaveTextContent("Test");
});

it("matches snapshot", () => {
  const tree = renderer.create(<CustomButton>Done</CustomButton>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshot of add task button", () => {
  const tree = renderer.create(<CustomButton>ADD TASK</CustomButton>).toJSON();
  expect(tree).toMatchSnapshot();
});
