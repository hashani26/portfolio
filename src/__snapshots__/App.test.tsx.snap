// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  className="ant-btn"
  data-testid="button"
  onClick={[Function]}
  type="button"
>
  <span>
    Done
  </span>
</button>
`;

exports[`matches snapshot of add task button 1`] = `
<button
  className="ant-btn"
  data-testid="button"
  onClick={[Function]}
  type="button"
>
  <span>
    ADD TASK
  </span>
</button>
`;
